numcodecs/compat.py

```diff
@@ -14,29 +14,47 @@
 
 if PY2:  # pragma: py3 no cover
 
+    buffer = buffer
     text_type = unicode
     binary_type = str
     integer_types = (int, long)
     reduce = reduce
 
 else:  # pragma: py2 no cover
 
+    buffer = memoryview
     text_type = str
     binary_type = bytes
     integer_types = int,
     from functools import reduce
 
 
+def to_buffer(v):
+    """Obtain a `buffer` or `memoryview` for `v`."""
+    b = v
+    if isinstance(b, np.ndarray):
+        if b.dtype.kind is 'O':
+            raise ValueError('cannot encode object array')
+        b = b.reshape(-1, order='A')
+        if b.dtype.kind in 'Mm':
+            b = b.view(np.uint8)
+        b = buffer(b)
+    elif PY2:  # pragma: py3 no cover
+        if not isinstance(b, array.array) and memoryview(b).format is 'O':
+            raise ValueError('cannot encode object array')
+        b = buffer(b)
+    else:      # pragma: py2 no cover
+        b = buffer(b)
+        if b.format is 'O':
+            raise ValueError('cannot encode object array')
+        b = b.cast('B').cast(b.format)
```

Comment:
Do we need to cast here? What does this add?

Reply:
It flattens the data while preserving the type. This [SO answer]( https://stackoverflow.com/a/44486410 ) explains it pretty well.

Ideally we would just do `b.cast(b.format)`, but `memoryview`s seem to want to either start or end in some kind of `char` type unfortunately. So we pass through `unsigned char`.